game/aamp_v2.ksy

```diff
@@ -0,0 +1,217 @@
+    doc-ref: https://zeldamods.org/wiki/AAMP#Parameter_list
+  - id: parameter_objects
+    type: parameter_object
+    repeat: expr
+    repeat-expr: header.num_objects
+    doc-ref: https://zeldamods.org/wiki/AAMP#Parameter_object
+  - id: parameters
+    type: parameter
+    repeat: expr
+    repeat-expr: header.num_parameters
+    doc-ref: https://zeldamods.org/wiki/AAMP#Parameter
+  - id: data_section
+    size: header.data_section_size
+    doc-ref: https://zeldamods.org/wiki/AAMP#Section_order
+  - id: string_section
+    size: header.string_section_size
+    doc-ref: https://zeldamods.org/wiki/AAMP#Section_order
+  - id: unknown_uint32_section
+    type: u4
+    repeat: expr
+    repeat-expr: header.num_unknown_uints
+    doc: Purpose is unknown. May be unused.
+    doc-ref: https://zeldamods.org/wiki/AAMP#Section_order
+types:
+  header:
+    seq:
+    - id: magic
+      contents: AAMP
+    - id: version
+      type: u4
+      doc: Should be "2"
+    - id: flags
+      type: u4
+      doc: |
+        TODO
+        Flags (LittleEndian 1 << 0, UTF8 1 << 1)
+        What does this mean?
+      doc-ref: https://zeldamods.org/wiki/AAMP#Header
+    - id: file_size
+      type: u4
+    - id: pio_verision
+      type: u4
+    - id: pio_offset
+      type: u4
+    - id: num_lists
+      type: u4
+    - id: num_objects
+      type: u4
+    - id: num_parameters
+      type: u4
+    - id: data_section_size
+      type: u4
+    - id: string_section_size
+      type: u4
+    - id: num_unknown_uints
+      size: 4
+      doc: number of uint32s after the string section
+    - id: pio_type
+      type: strz
+      size: 4
+      encoding: utf-8
+      doc: typically "xml"
+  parameter_list:
+    seq:
+      - id: name_crc32
+        type: u4
+      - id: child_lists
+        type: u4
+      - id: child_objects
+        type: u4
+    instances:
+      child_lists_offset:
+        value: child_lists & 16
+        doc: Offset to child lists, divided by 4 and relative to parameter list start
+      num_child_lists:
+        value: child_lists >> 16
+        doc: Number of child lists
+      child_objects_offset:
+        value: child_objects & 16
+        doc: Offset to child objects, divided by 4 and relative to parameter list start
+      num_child_objects:
+        value: child_objects >> 16
+        doc: Number of child objects
+  parameter_object:
+    seq:
+      - id: name_crc32
+        type: u4
+      - id: data
+        type: u4
+    instances:
+      child_offset:
+        value: data & 16
+      num_child_params:
+        value: data >> 16
+  parameter:
+    seq:
+      - id: name_crc32
+        type: u4
+      - id: data
+        type: u4
+    instances:
+      data_offset:
+        value: data & 24
+        doc:  Offset to data, divided by 4 and relative to parameter start.
+      parameter_type:
+        value: data >> 24
+        enum: parameter_type
+        doc-ref: https://zeldamods.org/wiki/AAMP#ParameterType
```

Comment:
This seems like a fine solution, but doesn't work in practice. Any idea why?
(Sorry if it's obvious, I'm still waking up.)
Current code:
```ksy
  parameter:
    seq:
      - id: name_crc32
        type: u4
        doc: |
          CRC32 checksum of the name of this parameter.
          Can be compared against this list to get the name:
        doc-ref: [SNIPPED]
      - id: data_offset
        type: b24
        doc:  Offset to data, divided by 4 and relative to parameter start.
      - id: parameter_type
        type: b8
        enum: parameter_type
        doc-ref: https://zeldamods.org/wiki/AAMP#ParameterType
```

Reply:
> This seems like a fine solution, but doesn't work in practice. Any idea why?

Well, let's see how the bytes are laid out. `data` has the little-endian byte order, so extracting the members `data_offset` and `parameter_type` works with the `d[3] d[2] d[1] d[0]` array:

```
                  d[3]                                      d[2]                                      d[1]                                      d[0]
  7    6    5    4    3    2    1    0      7    6    5    4    3    2    1    0      7    6    5    4    3    2    1    0      7    6    5    4    3    2    1    0
  p7   p6   p5   p4   p3   p2   p1   p0     f23  f22  f21  f20  f19  f18  f17  f16    f15  f14  f13  f12  f11  f10  f9   f8     f7   f6   f5   f4   f3   f2   f1   f0
```

_`data_offset` would be obtained as `0b{f23}{f22}{f21}...{f1}{f0}`, analogously `parameter_type` is `0b{p7}{p6}...{p1}{p0}`._

But note that the bytes are presented in little-endian order in the above diagram, so let's swap them back to the original order as present in the stream:

```
                  d[0]                                      d[1]                                      d[2]                                      d[3]
  7    6    5    4    3    2    1    0      7    6    5    4    3    2    1    0      7    6    5    4    3    2    1    0      7    6    5    4    3    2    1    0
  f7   f6   f5   f4   f3   f2   f1   f0     f15  f14  f13  f12  f11  f10  f9   f8     f23  f22  f21  f20  f19  f18  f17  f16    p7   p6   p5   p4   p3   p2   p1   p0
```
<p align=center><b>Diagram A ↑</b></p>

So yeah, true, it becomes apparent that `data_offset` cannot be parsed using the big-endian bit-sized integer parsing method described at https://doc.kaitai.io/user_guide.html#bit-ints-be, because it uses the following parsing direction:

```
               d[0]                              d[1]
  7   6   5   4   3   2   1   0     7   6   5   4   3   2   1   0     7   ...
  f7  f6  f5  f4  f3  f2  f1  f0    f15 f14 f13 f12 f11 f10 f9  f8    f7  ...
 ┃ ───────────────────────────> │  │ ───────────────────────────> │  │ ── ... ──>
        parsing direction       ╷  ↑                              ╷  ↑
                                └┄┄┘                              └┄┄┘
```

Which is pretty useless for reading the `data_offset`, because by applying this direction to **Diagram A** you will get `0b{f7}{f6}{f5}...{f0}{f15}{f14}...{f8}{f23}{f22}...{f16}`, and that's certainly not what you want.

So you need to use the [little-endian direction](https://doc.kaitai.io/user_guide.html#bit-ints-le) (indicated by `bit-endian: le` or `type: bXle`), which looks like this:

```
               d[0]                              d[1]
  7   6   5   4   3   2   1   0     7   6   5   4   3   2   1   0     7   ...
  f7  f6  f5  f4  f3  f2  f1  f0    f15 f14 f13 f12 f11 f10 f9  f8       ...
 │ <─────────────────────────── ┃  │ <─────────────────────────── │  │
 ╷       parsing direction                                        ↑
 └┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈>┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┘
```

Using this method on **Diagram A** will finally give you `0b{f23}{f22}{f21}...{f2}{f1}{f0}` (you can imagine it like the first parsed bit is the least significant bit `{f0}` from `d[0].0`, the second is `{f1}` from `d[0].1`, ..., the 8th is `{f7}` from `d[0].7`, the 9th is `{f8}` from `d[1].0` and so on).

---

In short, this should work:

```ksy
meta:
  bit-endian: le
# ...
types:
  parameter:
    seq:
      # ...
      - id: data_offset
        type: b24
        doc: Offset to data, divided by 4 and relative to parameter start.
      - id: parameter_type
        type: b8
        enum: parameter_type
        doc-ref: https://zeldamods.org/wiki/AAMP#ParameterType
```